File: tool4_junit/codeFiles/python/Q4_stack.py
import unittest

class Stack:
    def __init__(self, size):
        self.max_size = size
        self.stack_array = [0] * self.max_size
        self.top = -1

    def push(self, value):
        if self.top < self.max_size - 1:
            self.top += 1
            self.stack_array[self.top] = value
            print("Pushed:", value)
        else:
            print("Stack Overflow. Cannot push", value)

    def pop(self):
        if self.top >= 0:
            popped_value = self.stack_array[self.top]
            self.top -= 1
            print("Popped:", popped_value)
            return popped_value
        else:
            print("Stack Underflow. Cannot pop")
            return -1  
        
class TestStack(unittest.TestCase):
    def setUp(self):
        self.st = Stack(5)

    def test_push_pop(self):
        self.st.push(6)
        self.st.push(8)
        self.assertEqual(8, self.st.pop())
        self.assertEqual(6, self.st.pop())

if __name__ == '__main__':
    unittest.main()

File: tool4_junit/codeFiles/junit/src/junit/Q4_Stack.java
package junit;

public class Q4_Stack {
	private int maxSize;
    private int top;
    private int[] stackArray;

    public Q4_Stack(int size) {
        maxSize = size;
        stackArray = new int[maxSize];
        top = -1;
    }

    public void push(int value) {
        if (top < maxSize - 1) {
            stackArray[++top] = value;
            System.out.println("Pushed: " + value);
        } else {
            System.out.println("Stack Overflow. Cannot push " + value);
        }
    }

    public int pop() {
        if (top >= 0) {
            int poppedValue = stackArray[top--];
            System.out.println("Popped: " + poppedValue);
            return poppedValue;
        } else {
            System.out.println("Stack Underflow. Cannot pop");
            return -1; // Return a sentinel value or throw an exception as per your requirement
        }
    }
    
}
